sklearn/linear_model/base.py

```diff
@@ -195,13 +195,21 @@ def _preprocess_data(X, y, fit_intercept, normalize=False, copy=True,
                 X, X_scale = f_normalize(X, axis=0, copy=False,
                                          return_norm=True)
             else:
-                X_scale = np.ones(X.shape[1])
+                X_scale = np.ones(X.shape[1], dtype=X.dtype)
         y_offset = np.average(y, axis=0, weights=sample_weight)
+        if X.dtype == np.float32:
+            y_offset = np.float32(y_offset)
         y = y - y_offset
     else:
-        X_offset = np.zeros(X.shape[1])
-        X_scale = np.ones(X.shape[1])
-        y_offset = 0. if y.ndim == 1 else np.zeros(y.shape[1], dtype=X.dtype)
+        X_offset = np.zeros(X.shape[1], dtype=X.dtype)
+        X_scale = np.ones(X.shape[1], dtype=X.dtype)
+        if y.ndim == 1:
+            if X.dtype == np.float32:
+                y_offset = np.float32(0)
+            else:
+                y_offset = np.float64(0)
```

Comment:
What about replacing
`if X.dtype == np.float32:
    y_offset = np.float32(0)
else:
    y_offset = np.float64(0)`
above with just
`y_offset = X.dtype.type(0)`
?
Tested the dtype.type method with numpy 1.8.2 and 1.12.1
https://docs.scipy.org/doc/numpy/reference/generated/numpy.dtype.type.html